local-tests/chainrunner/chain.py

```diff
@@ -28,39 +29,41 @@ def __getitem__(self, i):
     def __iter__(self):
         return iter(self.nodes)
 
-    def bootstrap(self, binary, accounts, **kwargs):
-        """Bootstrap the chain. `accounts` should be a list of strings.
+    def bootstrap(self, binary, validator_accounts, accounts=None, **kwargs):
+        """Bootstrap the chain. `validator_accounts` should be a list of strings.
         Flags `--account-ids`, `--base-path` and `--raw` are added automatically.
         All other flags are taken from kwargs"""
+        accounts = accounts or []
         cmd = [check_file(binary),
                'bootstrap-chain',
                '--base-path', self.path,
-               '--account-ids', ','.join(accounts), '--raw']
+               '--account-ids', ','.join(validator_accounts), '--raw']
         cmd += flags_from_dict(kwargs)
 
         chainspec = op.join(self.path, 'chainspec.json')
         with open(chainspec, 'w', encoding='utf-8') as f:
             subprocess.run(cmd, stdout=f, check=True)
 
         self.nodes = []
-        for a in accounts:
+        for a in chain(validator_accounts, accounts):
             n = Node(binary, chainspec, op.join(self.path, a), self.path)
             n.flags['node-key-file'] = op.join(self.path, a, 'p2p_secret')
             self.nodes.append(n)
 
-    def set_flags(self, *args, **kwargs):
-        """Set common flags for all nodes.
+    def set_flags(self, *args, predicate=lambda n, i: True, **kwargs):
```

Comment:
I am thinking whether  `predicate` argument can be part of a `bootstrap`, as this is where `self.nodes` are constucted

Reply:
here I wanted something that could set different flags to different nodes but maybe it makes more sense to divide nodes as you have mentioned in the bootstrap and have 3 functions (set_flags, set_flags_for_validators, set_flags_for_nonvalidators). wdyt?

Reply:
Yes please good idea.